core/src/main/java/de/javakaffee/web/msm/storage/RedisStorageClient.java

```diff
@@ -0,0 +1,292 @@
+        
+        Callable<byte[]> callable = new RedisCommandCallable<byte[]>() {
+            @Override protected byte[] execute(BinaryJedis jedis) throws Exception {
+                return jedis.get(keyBytes(key));
+            }
+        };
+        
+        // Execute callable synchronously since we need to wait for the result anyway
+        try {
+            return callable.call();
+        }
+        catch (Exception e) {
+            if (e instanceof RuntimeException)
+                throw (RuntimeException)e;
+            else
+                throw new RuntimeException("Error getting key from Redis", e);
+        }
+    }
+    
+    @Override
+    public Future<Boolean> delete(final String key) {
+        if (_log.isDebugEnabled())
+            _log.debug(String.format("Deleting key in Redis (key=%s)", key));
+
+        return _executor.submit(new RedisCommandCallable<Boolean>() {
+            @Override protected Boolean execute(BinaryJedis jedis) throws Exception {
+                return jedis.del(keyBytes(key)) == 1;
+            }
+        });
+    }
+
+    @Override
+    public void shutdown() {
+        _pool.shutdown();
+    }
+    
+    private static int convertExp(int exp) {
+        if (exp <= 60*60*24*30) // thirty days
+            return exp;
+        else
+            return Math.max(exp - (int)(System.currentTimeMillis() / 1000), 1);
+    }
+    
+    private static byte[] keyBytes(String key) {
+        return key.getBytes(StandardCharsets.UTF_8);
+    }
+    
+    private abstract class RedisCommandCallable<T> implements Callable<T> {
+        @Override
+        public T call() throws Exception {
+            BinaryJedis jedis = null;
+
+            // Borrow an instance from Jedis without checking it for performance reasons and execute the command on it
+            try {
+                jedis = _pool.borrowInstance(false);
+                return execute(jedis);
+            } catch (JedisConnectionException e) {
+                // Connection error occurred with this Jedis connection, so now make sure to get a known-good one
+                // The old connection is not given back to the pool since it is defunct anyway
+                if (_log.isDebugEnabled())
+                    _log.debug("Connection error occurred, discarding Jedis connection: " + e.getMessage());
+
+                jedis = null;
+            } finally {
+                if (jedis != null)
+                    _pool.returnInstance(jedis);
+            }
+            
+            // Try to execute the command again with a known-good instance
+            try {
+                jedis = _pool.borrowInstance(true);
+                return execute(jedis);
+            } finally {
+                if (jedis != null)
+                    _pool.returnInstance(jedis);
+            }
+        }
+        
+        protected abstract T execute(BinaryJedis jedis) throws Exception;
+    }
+
+    private class JedisPool {
```

Comment:
The JedisPool depends on apache-commons-pool and I do not want this as a dependency